cmd/createtree/main_test.go

```diff
@@ -0,0 +1,228 @@
+	nonDefaultOpts.duplicatePolicy = nonDefaultTree.DuplicatePolicy.String()
+	nonDefaultOpts.displayName = nonDefaultTree.DisplayName
+	nonDefaultOpts.description = nonDefaultTree.Description
+
+	invalidEnumOpts := *validOpts
+	invalidEnumOpts.treeType = "LLAMA!"
+
+	invalidKeyTypeOpts := *validOpts
+	invalidKeyTypeOpts.privateKeyType = "LLAMA!!"
+
+	invalidPEMPath := *validOpts
+	invalidPEMPath.pemKeyPath = "/not/a/file"
+
+	emptyPEMPass := *validOpts
+	emptyPEMPass.pemKeyPass = ""
+
+	tests := []struct {
+		desc      string
+		opts      *createOpts
+		createErr error
+		wantErr   bool
+		wantTree  *trillian.Tree
+	}{
+		{
+			desc:     "validOpts",
+			opts:     validOpts,
+			wantTree: defaultTree,
+		},
+		{
+			desc:     "nonDefaultOpts",
+			opts:     &nonDefaultOpts,
+			wantTree: &nonDefaultTree,
+		},
+		{
+			// No mandatory opts provided
+			desc:    "defaultOptsOnly",
+			opts:    newOptsFromFlags(),
+			wantErr: true,
+		},
+		{
+			desc:    "invalidEnumOpts",
+			opts:    &invalidEnumOpts,
+			wantErr: true,
+		},
+		{
+			desc:    "invalidKeyTypeOpts",
+			opts:    &invalidKeyTypeOpts,
+			wantErr: true,
+		},
+		{
+			desc:    "invalidPEMPath",
+			opts:    &invalidPEMPath,
+			wantErr: true,
+		},
+		{
+			desc:    "emptyPEMPass",
+			opts:    &emptyPEMPass,
+			wantErr: true,
+		},
+		{
+			desc:      "createErr",
+			opts:      validOpts,
+			createErr: errors.New("create tree failed"),
+			wantErr:   true,
+		},
+	}
+
+	for _, test := range tests {
+		test := test
+		func() {
+			var out *string
+			println = func(a ...interface{}) (int, error) {
+				str := fmt.Sprintln(a...)
+				out = &str
+				return 0, nil
+			}
+			defer func() {
+				println = fmt.Println
+			}()
+
+			if test.createErr != nil {
+				server.err = test.createErr
+				defer func() {
+					server.err = nil
+				}()
+			}
+
+			tree, err := createTree(test.opts)
+			switch hasErr := err != nil; {
+			case hasErr != test.wantErr:
+				t.Errorf("%v: createTree() returned err = '%v', wantErr = %v", test.desc, err, test.wantErr)
+				return
+			case hasErr:
+				return
+			}
+
+			if diff := pretty.Compare(tree, test.wantTree); diff != "" {
+				t.Errorf("%v: post-createTree diff:\n%v", test.desc, diff)
+			}
+
+			if out == nil {
+				t.Errorf("%v: no output recorded", test.desc)
+			} else if want := fmt.Sprintf("%v\n", tree.TreeId); *out != want {
+				t.Errorf("%v: out = %v, want = %v", test.desc, *out, want)
+			}
+		}()
+	}
+}
+
+type fakeAdminServer struct {
+	err error
+}
+
+func startFakeServer() (*fakeAdminServer, net.Listener, func(), error) {
+	grpcServer := grpc.NewServer()
+	fakeServer := &fakeAdminServer{}
+	trillian.RegisterTrillianAdminServer(grpcServer, fakeServer)
+
+	lis, err := net.Listen("tcp", "")
+	if err != nil {
+		return nil, nil, nil, err
+	}
+	go grpcServer.Serve(lis)
+
+	stopFn := func() {
+		grpcServer.Stop()
+		lis.Close()
+	}
+	return fakeServer, lis, stopFn, nil
```

Comment:
Although it doesn't make much difference, I prefer returning the listener in this case. It's clearer than a string return.